### ISSUE TITLE
GPU.4 The runtime for NVScoreVariants, and whether the GPU pays

### ISSUE THREAD

**Issue by BenjaminDEMAILLE** [IPNP-BIPN/gatk-rs#1021]
Which Rust runtime executes the NVScoreVariants model, and whether a GPU path is worth building at all. #250 settled the numeric target; this settles the dependency.

## What the reference actually ships

`nvscorevariants.py` loads the model with `model = torch.load(model_file)`. That is a **pickled `nn.Module`**, not TorchScript and not a `state_dict`: unpickling it requires the `scorevariants.models` classes on `sys.path`. No Rust runtime can open it directly. Whatever we pick, a one-shot Python step extracts the weights to safetensors or npz, and that artefact is then pinned by hash like every other reference input.

The architecture is small. `GATK_CNN_1D` is 70 lines:

```
Conv1d(4 -> 256, k=12) -> ReLU
Conv1d(256 -> 256, k=12) -> ReLU -> Dropout(0.1)
Conv1d(256 -> 128, k=12) -> ReLU -> Dropout(0.1)
BatchNorm1d(7, eps=1e-3, momentum=0.99)
Linear(7 -> 40) -> ReLU
Linear(12200 -> 40) -> ReLU -> Dropout(0.2)
Linear(47 -> 4)
```

`gatk_2d.py` is 75 lines. In `eval()` the dropouts vanish, so the whole surface is conv1d, ReLU, batchnorm, linear and a concat.

## The candidates

| | tch-rs | candle | ferrotorch |
|---|---|---|---|
| stars | 5480 | ~18k | **5** |
| crate downloads | millions | millions | **662** |
| last push | 2026-08-23 | active | 2026-07-15 |
| contributors | many | many | 6 |
| first published | 2020 | 2023 | 2026-03 |
| GPU backends | CUDA, MPS | CUDA, Metal | **Intel XPU only** |
| native dependency | libtorch, ~200 MB | none | none |

`ferrotorch` is out. Five stars, six months old, and the only GPU backend published on crates.io is Intel XPU. Nothing for CUDA or Metal. Betting the port's ML surface on it means maintaining a deep learning framework alongside the port.

`tch-rs` runs the real libtorch kernels, which is the best available fidelity, and exposes `Device::Mps` and `Device::Cuda`. The cost is ~200 MB of libtorch shipped with the CLI and a strict version coupling between the crate and the library. It does not save the Python extraction step, because it cannot read the pickle either.

`candle` covers CUDA and Metal, stays pure Rust at build time, and reads safetensors natively. For seven layers it is the smaller commitment.

## The GPU question is not the runtime question

Per variant this model is three convolutions and three dense layers. The GPU only pays if variants are batched by the thousand, and the likely bottleneck is BAM/VCF reading and the one-hot encoding in `encoders.py`, not the convolution. The reference runs this on CPU without complaint.

Also: a GPU path cannot be the byte-identical path. #250 measured CPU against MPS at 4 ULP on 7 of 8 floats with the shipped model. Same conclusion as GPU.1: the reference path stays on CPU, single-threaded, and any GPU path is a separate fast mode that is not compared against the reference.

## What to measure before choosing

1. Encode a realistic batch of variants, time CPU inference alone against total tool wall time. If inference is under 10% the GPU path is a distribution cost with no return.
2. Extract the weights once, run the same batch through candle on CPU, and diff against the PyTorch 2.1.0 reference at the precision the VCF actually prints. That decides whether a pure-Rust reimplementation holds the byte, and therefore whether libtorch is needed at all.
3. Only if 2 fails does `tch` become necessary. The extraction from step 2 stays valid either way.

Parent: #33. Related: #73, #250.
